ibind/client/ibkr_client_mixins/oauth_mixin.py

```diff
@@ -0,0 +1,423 @@
+
+
+
+import configparser
+from dotenv import load_dotenv
+from Crypto.PublicKey import RSA
+from Crypto.Signature import PKCS1_v1_5 as PKCS1_v1_5_Signature
+from Crypto.Cipher import PKCS1_v1_5 as PKCS1_v1_5_Cipher
+from Crypto.Hash import SHA256, HMAC, SHA1
+from Crypto.IO import PEM
+from Crypto.PublicKey import DSA
+
+from cryptography.hazmat.primitives.asymmetric import dh
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.serialization import load_pem_parameters
+
+from Crypto.PublicKey import RSA
+from Crypto.Signature import PKCS1_v1_5 as PKCS1_v1_5_Signature
+from Crypto.Cipher import PKCS1_v1_5 as PKCS1_v1_5_Cipher
+from Crypto.Hash import SHA256, HMAC, SHA1
+from urllib import parse
+import string
+import random
+import base64
+import time
+
+class OAuthMixin():  # pragma: no cover
+
+    def req_live_session_token(self) -> tuple[str, int]:
+        """Get live session token and access token from IBKR Web API used to make API endpoint calls
+
+        Raises:
+            Exception: _description_
+            Exception: _description_
+
+        Returns:
+            tuple[token,token]: live access token, access token
+        """
+
+        REQUEST_URL = "oauth/live_session_token"
+        # REQUEST_URL = "https://api.ibkr.com/v1/api/oauth/live_session_token"
+        # REQUEST_METHOD = "POST"
+        ENCRYPTION_METHOD = "RSA-SHA256"
+        
+        load_dotenv()
+        config = configparser.ConfigParser()
+        config.read('D:\\git_repos\\oauth_env\\oauth_test.env')
+
+        dh_prime=self.pem_to_dh_prime(pem_file_path=config['ibkr']['DH_PRIME_FP'])
+
+        dh_random = self.generate_dh_random_bytes()
+        dh_challenge = self.generate_dh_challenge(
+            dh_prime=dh_prime,
+            dh_generator=int(config['ibkr']['DH_GENERATOR']),
+            dh_random=dh_random,
+        )
+        prepend = self.calculate_live_session_token_prepend(
+            access_token_secret=config['ibkr']['ACCESS_TOKEN_SECRET'],
+            private_encryption_key=self.read_private_key(private_key_fp=config['ibkr']['ENCRYPTION_KEY_FP']),
+        )
+
+        extra_headers={"diffie_hellman_challenge": dh_challenge}
+
+        # get OAuth headers - move this to the rest_client.py request function?? but then many variables to pass or make global or class attributes
+        # header_oauth=self.get_headers(
+        #     request_method=REQUEST_METHOD,
+        #     request_url=REQUEST_URL,
+        #     oauth_token=config['ibkr']['ACCESS_TOKEN'],
+        #     signature_method=ENCRYPTION_METHOD,
+        #     extra_headers={"diffie_hellman_challenge": dh_challenge},
+        #     prepend=prepend,
+        #     )
+
+        # call ibind request
+        result=self.post(path=REQUEST_URL, log = True,signature_method=ENCRYPTION_METHOD,extra_headers=extra_headers,prepend=prepend)
```

Comment:
`post` doesn't have signature_method, extra_headers or prepend parameters, so I'm assuming you're passing these as **kwargs. But then, these never get used. What is the reason to pass these? Do you pass them just so that they could be found in the returned Result object?

Reply:
Yes passed as **kwargs.
```signature_method, extra_headers and prepend``` are used to get the live_session_token, but once this has been obtained, they are not needed for other endpoint calls, so I passed them as **kwargs.
But this conflicts with other endpoint calls which need **kwargs other than OAuth variables. 
This method needs to be improved. Any ideas?
```
# add IBKR OAuth headers to request function
response = requests.request(method,url,headers=header_oauth,params=params,timeout=10)
# response = requests.request(method, url, verify=self.cacert,headers=header_oauth,params=params, timeout=self._timeout, **kwargs)
```

Reply:
This is where the oauth headers are obtained, in rest_client.py

# get headers for OAuth call
        if self._use_oauth:

            endpoint = endpoint.lstrip("/")
            oauth_base_url='https://api.ibkr.com/v1/api/'
            url = f"{oauth_base_url}{endpoint}"

            header_oauth= self.get_headers(
                request_method=method,
                request_url=url,
                **kwargs
            )
        else:
            endpoint = endpoint.lstrip("/")
            url = f"{self.base_url}{endpoint}"

Reply:
Gotcha! I appreciate the thorough explanation. My first thought is, why don't we generate  `signature_method`, `extra_headers` and `prepend` in `IbkrClient.get_headers()` method? The code used to generate these doesn't rely on any input arguments other than the ones you read from the config, which we eventually will turn into environment variables - and in either case these can be read here just as much as they could be in the `IbkrClient.get_headers()` right?

Reply:
If we pass the url endpoint to ```get_headers``` and if ```url=oauth/live_session_token``` then get ```signature_method, extra_headers and prepend``` otherwise don't could be cleaner code?

So bundle function ```req_live_session_token``` into ```get_headers```, but this makes ```get_headers``` more complex. There is a fair amount happening in ```req_live_session_token function```.

Reply:
Maybe once all bugs and request either via localhost or OAuth are working we can revisit this? I think this is lower priority.

Reply:
I feel that there is a lurking bug somewhere in this section of code in ```rest_client```, like using **kwargs for OAuth and for endpoint calls.
```
for attempt in range(self._max_retries + 1):
            try:
                # add IBKR OAuth headers to request function
                response = requests.request(method,url,headers=header_oauth,params=params,timeout=10)
                # response = requests.request(method, url, verify=self.cacert,headers=header_oauth,params=params, timeout=self._timeout, **kwargs)
                result = Result(request={'url': url, **kwargs})
                return self._process_response(response, result)
```

Reply:
Hey @hughandersen I've submitted a PR to your branch in order to help solving this. Let me know there if it helps.